pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -658,13 +658,20 @@ func applyOutlierDetection(c *cluster.Cluster, outlier *networking.OutlierDetect
 
 	// Disable panic threshold by default as its not typically applicable in k8s environments
 	// with few pods per service.
-	// To do so, set the healthy_panic_threshold field even if its value is 0 (defaults to 50).
+	// To do so, set the healthy_panic_threshold field even if its value is 0 (defaults to 50 in Envoy).
 	// FIXME: we can't distinguish between it being unset or being explicitly set to 0
-	if outlier.MinHealthPercent >= 0 {
+	minHealthPercent := outlier.MinHealthPercent
+	if minHealthPercent >= 0 {
 		if c.CommonLbConfig == nil {
 			c.CommonLbConfig = &cluster.Cluster_CommonLbConfig{}
 		}
-		c.CommonLbConfig.HealthyPanicThreshold = &xdstype.Percent{Value: float64(outlier.MinHealthPercent)} // defaults to 50
+		// When we are sending unhealthy endpoints, we should disble Panic Threshold. Otherwise
+		// Envoy will send traffic to "Unready" pods when the percentage of healthy hosts fall
+		// below minimum health percentage.
+		if features.SendUnhealthyEndpoints {
```

Comment:
Should we always disable healthy panic threshold if SendUnhealthyEndpoints? Even if the user is explicitly enabling it in outlier detection?

I would rather have it default to disabled when SendUnhealthyEndpoints is on and no users' outlier detection is set, since it may be desired to do both things. Also would clarify users of the implications of turning SendUnhealthyEndpoints with healthy panic threshold.

Reply:
> I would rather have it default to disabled when SendUnhealthyEndpoints is on and no users' outlier detection is set, since it may be desired to do both things. Also would clarify users of the implications of turning SendUnhealthyEndpoints with healthy panic threshold.

But that would send traffic to unready pods which is against K8s contract as @howardjohn mentioned.  I would rather ignore this and clarify it in the docs of outlier detection. There is no good solution that works for both.

Reply:
I don't suppose envoy has support for 3 modes? Healthy, unhealthy (but can send in panic), and unhealthy (never send)?

that is the ideal

Reply:
No it is not there. We will have to enhance it

Reply:
Can we open an issue in envoy?